integration/testdata/fluentd-multiple-lockfiles.cdx.json.golden

```diff
@@ -51,7 +51,7 @@
       ]
     },
     {
-      "bom-ref": "pkg:deb/debian/adduser@3.118?arch=all\u0026distro=debian-10.2",
+      "bom-ref": "pkg:dpkg/adduser@3.118?arch=all",
```

Comment:
Looks like the PURL is now wrong. The type should be `deb`.
https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#deb

Reply:
This is sth that @DmitriyLewen pointed out on this [previous conversation](https://github.com/aquasecurity/trivy/pull/5439#discussion_r1396846978) but the problem is that we don't have the [Metadata](https://github.com/aquasecurity/trivy/blob/main/pkg/types/report.go#L26-L36) info in pkg (apk, rpm, etc.) analyzers.

After I moved the pURLs generation to the "analysis" part so `PkgIdentifier` is also part of the `Package` type per your suggestion, I was unable to find a way to obtain it. This is the reason why [NewPackageIdentifier](https://github.com/aquasecurity/trivy/blob/8141d33e38d8a6385d5a1eebd153749feeb8dbc6/pkg/purl/purl.go#L201-L214) isn't receiving any metadata argument.

I'd appreciate help on this because I don't really know how to achieve it.

Reply:
We can move logic to creating purl's for pkg managers (dpg, rpm, apk) to [handler](https://github.com/aquasecurity/trivy/tree/main/pkg/fanal/handler).

This will allow us to create simpler code (we build purl in 1 place for all pkgs, instead of collecting purl for each package separately).
Also we will have info about OS.

Reply:
@DmitriyLewen proposal implemented at https://github.com/aquasecurity/trivy/pull/5439/commits/015e9a5238f92b4afa02526bb6ab37b15e4c4812 

This proposal doesn't moves entirely the pkg identifier generation to handler but overwrites the work done by pkg managers when it's possible. This way, we also have pkg identifiers in distroless & similar images that may use pkg managers without an actual OS.

Reply:
I'd appreciate if you could re-generate amd64 golden images by the way.

Reply:
But this golden file still has wrong purl:
https://github.com/aquasecurity/trivy/blob/bc2e8d62c3871e2f7a7be7225d08e7c257db5197/integration/testdata/fluentd-multiple-lockfiles.cdx.json.golden#L54

Reply:
> I'd appreciate if you could re-generate amd64 golden images by the way.

I will do that after we finish with handler logic.

Reply:
Is the fluentd image used to generate `trivy/integration/testdata/fluentd-multiple-lockfiles.cdx.json.golden` distroless?

Reply:
Sorry for delay.
We use `testdata/fixtures/images/fluentd-multiple-lockfiles.tar.gz` image:
https://github.com/aquasecurity/trivy/blob/b5874e3ad38e77ac86eedd7a65785b2933f3685f/integration/client_server_test.go#L409-L416

Reply:
I've trying to figure out why it uses `pkg:dpkg` instead of `pkg:deb/debian`. To do so, I introduced the log below in my branch:

```diff
diff --git a/pkg/fanal/handler/ospkgid/ospkgid.go b/pkg/fanal/handler/ospkgid/ospkgid.go
index 991abadb..7beb0db1 100644
--- a/pkg/fanal/handler/ospkgid/ospkgid.go
+++ b/pkg/fanal/handler/ospkgid/ospkgid.go
@@ -27,6 +29,7 @@ func newSystemPackagesPostHandler(artifact.Option) (handler.PostHandler, error)
 // metadata info since they were generated in pkg (apk, rpm, etc.) analyzers
 func (h systemPackagesPostHandler) Handle(_ context.Context, _ *analyzer.AnalysisResult, blob *types.BlobInfo) error {
        if blob != nil && len(blob.PackageInfos) > 0 {
+               log.Logger.Info(fmt.Sprintf("Blob with %s diff ID has OS family: %s", blob.DiffID, blob.OS.Family))
                blob.PackageInfos = overwritePkgIdentifiers(blob.PackageInfos, blob.OS)
        }
```

As we can see below:

```bash
INFO	Blob with sha256:831c5620387fb9efec59fc82a42b948546c6be601e3ab34a87108ecf852aa15f diff ID has OS family: debian
INFO	Blob with sha256:75e43d55939745950bc3f8fad56c5834617c4339f0f54755e69a0dd5372624e9 diff ID has OS family:
```

The scanner only recognizes OS family in one of the layers containing packages info. Therefore, the packages detected in the other layer are not adapted with the right PURL. I need further advice on how to address this.

Reply:
Thanks @DmitriyLewen ! This inspired me to address the issue in this commit: https://github.com/aquasecurity/trivy/pull/5439/commits/052555ccc0a0d2bd9393779861ba08327dabec2e